Userland/Games/ColorLines/MarbleBoard.h

```diff
@@ -0,0 +1,356 @@
+#include <AK/HashTable.h>
+#include <AK/IterationDecision.h>
+#include <AK/NumericLimits.h>
+#include <AK/Queue.h>
+#include <AK/Random.h>
+#include <AK/Vector.h>
+#include <LibGfx/Point.h>
+#include <LibGfx/Size.h>
+
+class MarbleBoard final {
+public:
+    using Color = Marble::Color;
+    using Point = Gfx::IntPoint;
+    using PointArray = Vector<Point>;
+    using SelectedMarble = Marble;
+    using PreviewMarble = Marble;
+    using MarbleArray = Vector<Marble>;
+
+    static constexpr Gfx::IntSize board_size { 9, 9 };
+    static constexpr size_t number_of_preview_marbles = 3;
+    static constexpr Color empty_cell = Marble::empty_cell;
+
+    using PreviewMarbles = Array<PreviewMarble, number_of_preview_marbles>;
+
+    MarbleBoard()
+    {
+        reset();
+    }
+
+    ~MarbleBoard() = default;
+
+    MarbleBoard(MarbleBoard const&) = delete;
+
+    [[nodiscard]] bool has_empty_cells() const
+    {
+        bool result = false;
+        for_each_cell([&](Point point) {
+            result = is_empty_cell_at(point);
+            return result ? IterationDecision::Break : IterationDecision::Continue;
+        });
+        return result;
+    }
+
+    [[nodiscard]] PointArray get_empty_cells() const
+    {
+        PointArray result;
+        for_each_cell([&](Point point) {
+            if (is_empty_cell_at(point))
+                result.append(point);
+            return IterationDecision::Continue;
+        });
+        random_shuffle(result);
+        return result;
+    }
+
+    void set_preview_marble(size_t i, PreviewMarble const& marble)
+    {
+        VERIFY(i < number_of_preview_marbles);
+        m_preview_marbles[i] = marble;
+    }
+
+    [[nodiscard]] bool place_preview_marbles_on_board()
+    {
+        if (!ensure_all_preview_marbles_are_on_empty_cells())
+            return false;
+        for (auto const& marble : m_preview_marbles)
+            if (!place_preview_marble_on_board(marble))
+                return false;
+        return true;
+    }
+
+    [[nodiscard]] bool check_preview_marbles_are_valid()
+    {
+        // Check marbles pairwise and also check the board cell under this marble is empty
+        return m_preview_marbles[0].position() != m_preview_marbles[1].position() && m_preview_marbles[0].position() != m_preview_marbles[2].position()
+            && m_preview_marbles[1].position() != m_preview_marbles[2].position()
+            && is_empty_cell_at(m_preview_marbles[0].position())
+            && is_empty_cell_at(m_preview_marbles[1].position())
+            && is_empty_cell_at(m_preview_marbles[2].position());
+    }
+
+    [[nodiscard]] bool update_preview_marbles(bool use_current)
+    {
+        auto empty_cells = get_empty_cells();
+        for (size_t i = 0; i < number_of_preview_marbles; ++i) {
+            auto marble = m_preview_marbles[i];
+            // Check marbles pairwise and also check the board cell under this marble is empty
+            auto const is_valid_marble = [&]() {
```

Comment:
This logic attempts to preserve the preview marbles that were generated in previos step. I.e. when the user moves the existing marble over one of the preview marbles, and it does not cause marble line removal, we need to update all the preview marbles to have 3 non-overlapping marbles to put them on board. One of the marbles must be regenerated (because it is now below the user-moved marble), and we need to preserve the other marble locations/colors. Also, we need to ensure that preview marbles do not overlap with each other and occupy empty cells on board.